### ISSUE TITLE
Separating mongo, ddp, and accounts-base

### ISSUE THREAD

**Issue by ccorcos** [meteor/meteor#4915]
**_3 Upvotes**_ These three packages are surprisingly intertwined:

`ddp` handles `Meteor.methods`, `Meteor.publish` and `Meteor.subscribe`. `Meteor.publish` and `Meteor.subscribe` assume the existence of `mongo` and `minimongo` to move data from server to client. `accounts-base` assumes the existence of `mongo` and `minimongo` to store user records on the client and the server. And lastly, `ddp` gives you `this.userId` within `Meteor.publish` and `Meteor.methods` which involves everything above.

After building `ccorcos:any-db`, I realized that for Meteor to really support other databases well, there needs to be some more fundamental changes to the underlying core packages. The package names are entirely up for debate, but here's what I want to propose:
- ddp-methods
- ddp-pub
- ddp-sub
- accounts-generic
- ddp-sub-mongo
- accounts-mongo
## `ddp-methods`

It seems to me like `Meteor.methods` is a completely separate functionality from `Meteor.publish` and `Meteor.subscribe` so I think it would be reasonable to pull this out into its own package. Also, some people like @msavin have been blogging about how to make Meteor less reactive so it would be cool to have a lean Meteor app without pub-sub if you wanted.
## `ddp-pub`

For databases with reactive drivers like `mongo` that support `Cursor.observeChanges`, then we can publish the the same way we're used to. However, for databases that aren't reactive, we can use the `diff-sequence` package to poll and diff the query results. And if we cache these subscriptions based on their query, we can proactively trigger a refresh whenever the data is changed. This is what I did in `ccorcos:any-db` and its actually really simple. The only thing missing here is [DDP doesn't currently support ordered publications](https://github.com/meteor/meteor/blob/devel/packages/ddp/DDP.md#procedure-2). I built [a hack for `ccorcos:any-db` to send ordered collections](https://github.com/ccorcos/meteor-any-db/blob/master/src/db.coffee#L149), but this is not ideal.

I'm imagining something that looks more like this:

``` js
Meteor.publish('messages', {ordered:true, cursor:false}, function(roomId) {
 return Messages.find({roomId:roomId}).fetch()
})
```

And in a `Meteor.method`, you can trigger the subscription to be refreshed on a write.

``` js
Meteor.methods({ 
  addMsg:function(roomId, text){ 
    if (Meteor.isServer) {
      Messages.insert({roomId:roomId, text:text})
      Meteor.update('messages', roomId)
    }
  } 
});
```
## `ddp-sub-mongo`

`ddp-sub-mongo` is exactly the same functionality we have right now with minimongo. All data is thrown into the appropriate `minimongo` collection and we can query the collection to get results out. This works great most of the time, but there a two very related pain points that I want to address:

1) It can be impossible to separate which data belongs to which subscription on the client. For example:

``` js
Products = new Mongo.Collection('products')
Meteor.subscribe('bestProducts')
Meteor.subscribe('worstProducts')
Products.findOne() // => {_id:'1234', name: 'Crayola Colored Bubbles'}
```

2) Some complicated database queries can sort data based on information that isn't available on the client, thus it can be impossible to sort the data appropriately.

These two issues can be overcome by being able to publish ordered collections and keeping track of which documents belong to which (possibly multiple) subscriptions -- which leads me to...
## `ddp-sub`

This package allows you to subscribe to ordered collections and keeps your subscriptions separate from each other.

Here's a `ddp-sub` example with Blaze:

``` jsx
Template.chatroom.onRendered(function() {
 this.autorun(() => {
   this.messages = Meteor.subscribe('messages', this.currentData().roomId)
 })
})
Template.chatroom.helpers({
 messages: function() {
   return Template.instance().messages.fetch()
 }
})
```

Here's a `ddp-sub` example with React:

``` jsx
React.createClass({
 mixins: [MeteorDataMixin],
 getMeteorData: function() {
   const messages = Meteor.subscribe('messages', this.props.roomId)
   return {messages: messages.fetch()}
 },
 // ...
})
```
## `accounts-generic`

I don't know a lot about how the accounts package works so I could be wrong, but it seems to me like a generic accounts package could expose an API to use any storage method on the client and the server.

``` js
// on the server
Users = new Mongo.Collection('custom-users')
Accounts.set(function(record) {
 Users.upsert({_id: record._id}, record)
})
Accounts.get(function(userId) {
 Users.findOne(userId)
})
// on the client
Accounts.set(function(cookie) {
 saveCookie(cookie)
})
Accounts.get(function() {
 return getCookie()
})
```

The ddp packages should be able to use these methods to authenticate users and provide `this.userId` in `Meteor.methods` and `Meteor.publish`. In addition to being able to use any storage method you want, I think devs with existing stacks would be much more open to adopting Meteor if they didn't have to migrate their entire user database to use Meteor's accounts packages.
## Implications
- Using `ddp-pub-mongo` and `accounts-mongo`, we should get the same exact functionality we have right now.
- It should possible to user `mongo` on the server without `minimongo` on the client using `ddp-sub` instead of `ddp-sub-mongo`.
### P.S. Latency Compensation

This is the only ugly part of all of this. Mongo and Minimongo do a great job at latency compensation and reactivity. Its a huge feat, but it doesnt seem likely to happen for other databases anytime soon. I dont have a perfect solution here but I have something that works.

Basically, I buffer the subscription data with a ReactiveVar:

``` jsx
var messages = new ReactiveVar([])
Tracker.autorun(() => {
  messagesSub = Meteor.subscribe('messages', roomId)
  Tracker.autorun(() => {
    messages.set(this.messagesSub.fetch())
  })
})
```

Then, I can latency compensate with the ReactiveVar however I want:

``` jsx
function submitMsg(roomId, text) {
  var prevMsgs = messages.get()
  messages.set([{roomId, text, unverified:true}].concat(prevMsgs))
  Meteor.call('addMsg', roomId, text, function(err) {
    if (err) {
      messages.set(prevMsgs)
    }
    // without an error, the subscription should update the messages
  })
}
```

---

Anyways, I just wanted to get a discussion started here. What are you thoughts?

**Comment by laosb** [comment]
+1 for this.

**Comment by zimme** [comment]
One thing I noticed with this is that in these examples the `subscribe` call doesn't seem to return a subscription handle but the cursor/data returned by the publish function, right?

**Comment by ccorcos** [comment]
Subscribe returns both! You can fetch and stop a subscription from what is returned. I've actually used this object to handle paging as well.

**Comment by zimme** [comment]
Oh, I've actually missed something :open_mouth:

That's undocumented too, so that might be good to fix also.

**Comment by rclai** [comment]
What? You can call `.fetch()` on the current `Meteor.subscribe`?

**Comment by zimme** [comment]
I just tried and I can't, maybe I missed something?

**Comment by rclai** [comment]
Oh then it must be theoretical code from his proposal.

**Comment by ccorcos** [comment]
Haha yes, precisely. That will allow you to distinguish what data belongs to which subscriptions.

**Comment by stubailo** [comment]
Thanks for the feature request! We welcome discussions about how to make Meteor better. If you haven't yet, check out our wiki page about [feature requests on GitHub](https://github.com/meteor/meteor/wiki/Contributing-to-Meteor#feature-requests).

**Comment by stubailo** [comment]
@ccorcos I think the best road to getting some of this stuff started would be to propose some small incremental improvements to decouple accounts from Mongo, or some other part. It is hard to have the discussion about everything at once.

**Comment by ccorcos** [comment]
@stubailo, definitely. I just didn't want to overwhelm. Here's some more thoughts with code examples:

### ddp-methods

I could be wrong, but it looks like `ddp-common` ([1])  is basically this exact package.

### ddp-pub and ddp-sub

- publications need to specify which subscriptionId they belong to. there is a subscription id being passed around between client and server:

```coffee
Meteor.publish name, ->
  subId = this._subscriptionId
  # ...

handle = Meteor.subscribe name, ->
  # ...

subId = handle.subscriptionId
```

But the problem is that this id is unavailable when parsing ddp data messages on the client:

```coffee
Meteor.connection.registerStore 'any-db',
  update: (msg) ->
    # what the subId?
```

- publications also need to handle ordered collections.

This means that we need to be able to send `addedBefore` and `movedBefore` messages over ddp. Looks like that will involve modifying the way we represent client sessions on the server ([2], [3]).

Right now, with `any-db` I've implemented a hack to work around both of these previous issues by creating a special key for every document that is published which specifies the subscription it belongs to and the position if its an ordered collection. For example:

```coffee
publishCursorOrdered = (pub, cursor, subId) ->
  cursor.observeChanges
    addedBefore: (id, fields, before) ->
      fields['any-db'] = {}
      # we have to add a random string so merge-box sees this as a new value
      fields['any-db'][subId] = Random.hexString(10) + '.' + before.toString()
      pub.added('any-db', id, fields)
    # ...
    

Meteor.connection.registerStore 'any-db',
  update: (msg) ->
    if msg.fields['any-db'] # for any addedBefore or movedBefore messages...
      for subId, value of msg.fields['any-db']
        if value is undefined
          # merge-box sends undefined if the field is cleared from the publication,
          # meaning remove this document from this subId.
        else
          before = value.split('.')[1]
          if before is "null" then before = null
          # now we know the subId and the position in that subscription
```

As I said, this is a hack but I think ordered publications ought to be supported. `diff-sequence` supports it do most of the work is already done.

- publications need to handle non-cursors. 

This is pretty easy using the `diff-sequence` package.

```coffee
publishList = (name, fetcher) ->
  Meteor.publish name, (query) ->
    pub = this
    docs = []
    fetch = fetcher.bind(null, query)
    refresh = ->
      newDocs = fetch()
      ordered = true
      DiffSequence.diffQueryChanges(
        ordered,
        docs,
        newDocs,
        {
          addedBefore: (id, fields, before) ->
            # ...
      })
      docs = newDocs
    refresh()
    pub.ready()
    key = JSON.stringify([name, query])
    unless pubs[key]
      pubs[key] = {}
    pubs[key][pub.subscriptionId] = refresh
```

Now, this `refresh` function can be saved somewhere to trigger the subscription to be refreshed on a database write. There also a good opportunity for caching here -- if there are 50 connections to clients that are all subscribed to the same thing, theres no need to fetch the same data from a remote database 50 times on every refresh.


### accounts-generic

This is something I have much less experience with. I tried digging through the code with little success. Looks like `Accounts.get` could be substituted here ([4]). `Accounts.set` on the server could be used here ([5]) and on the client here ([6]). The mongo collection may be directly accessed at different points in this package, but those ought to be replaced by an api method. Using `connection.setUserId` is a pretty nice API already to set the logged in user ([7]), so it looks like user accounts are already family decoupled from ddp. :)

### Summary

- It looks like `accounts-generic` wouldn't be too hard to build. We just need to go through `accounts-base` and create an API for setting and getting wherever Mongo or local storage is used. 
- Supporting ordered and unordered publications will require a decent amount of work -- DDP (including merge-box) needs to support `addedBefore` and `movedBefore` messages which send the position and the subscriptionId with each message.
- Caching and refreshing non-cursor publications needs some more thought. If 1000 users are connected to the same publication with the same exact query, then we don't need to fetch and diff 1000 times.

[1]: https://github.com/meteor/meteor/tree/devel/packages/ddp-common
[2]: https://github.com/meteor/meteor/blob/devel/packages/ddp-server/livedata_server.js#L119
[3]: https://github.com/meteor/meteor/blob/devel/packages/ddp-server/livedata_server.js#L332
[4]: https://github.com/meteor/meteor/blob/devel/packages/accounts-base/accounts_common.js#L47
[5]: https://github.com/meteor/meteor/blob/devel/packages/accounts-base/accounts_server.js#L1191
[6]: https://github.com/meteor/meteor/blob/devel/packages/accounts-base/localstorage_token.js#L43
[7]: https://github.com/meteor/meteor/blob/devel/packages/accounts-base/localstorage_token.js#L125

**Comment by ccorcos** [comment]
@stubailo, let me know your thoughts and where I ought to start.

**Comment by samcorcos** [comment]
+1

**Comment by Kestanous** [comment]
+1 with two additional thoughts:
1.  It would be really nice if subscriptions could return total count and current count of all its own docs. That is to say the total number of documents the server found and plans to send down, and the current number of docs the client has received. This would let us build much more accurate loading screens. 

2. Almost the revise of point 1, but with methods. Most likely used with offline apps, but not necessary. The number of insert/updates/outstanding methods the client has vs what the server has processed. This will allow us to tell users if all data has synced and it is safe to close the app. (recognize this may be vastly more extensive and may not be best in core, at least until offline has proper support)

**Comment by samcorcos** [comment]
@Kestanous I like those suggestions! Good call.

**Comment by ccorcos** [comment]
1. I love an accurate loading bar. haha. This would require good amount of work

Inside `Meteor.publish` you can call `this.added` ([1][1]) which, after diffing with the documents belonging to the client's Session, ends up sending a DDP message here ([2][2]). Then on the client, you receive a message here ([3][3]) and then it calls an update ([4][4]) on a store (from the example above: `Meteor.connection.registerStore`). Thats a ton of work though.

An easier way to do it, though kind of a hack, would be to publish each (5) document with an extra field:

```js
var cursor = Posts.find()
var total = cursor.count()
var i = 0
cursor.observeChanges({
    added: function (id, fields) {
      fields.progress = [i, total]
      sub.added(collection, id, fields);
    },
    changed: function (id, fields) {
      sub.changed(collection, id, fields);
    },
    removed: function (id) {
      sub.removed(collection, id);
    }
})
```

[1]: https://github.com/meteor/meteor/blob/devel/packages/ddp-server/livedata_server.js#L1204
[2]: https://github.com/meteor/meteor/blob/devel/packages/ddp-server/livedata_server.js#L332
[3]: https://github.com/meteor/meteor/blob/devel/packages/ddp-client/livedata_connection.js#L1181
[4]: https://github.com/meteor/meteor/blob/devel/packages/ddp-client/livedata_connection.js#L1226
[5]: https://github.com/meteor/meteor/blob/devel/packages/mongo/collection.js#L302

**Comment by ccorcos** [comment]
2. I would suggest using Meteor.methods as opposed to insert, update, remove with minimongo. But then you could create your own manager abstraction to count method calls, and determine what methods are still pending a callback.

**Comment by hwillson** [comment]
To help provide a more clear separation between feature requests and bugs, and to help clean up the feature request backlog, Meteor feature requests are now being managed under the https://github.com/meteor/meteor-feature-requests repository. 

This feature request will be closed here, but anyone interested in migrating this feature request to the new repository (to make sure it stays active), can click [here](https://github.com/meteor/meteor-feature-requests/issues/new?title=Separating%20mongo%2C%20ddp%2C%20and%20accounts-base&body=Migrated%20from:%20meteor/meteor%234915) to start the feature request migration process. This manual migration process is intended to help identify which of the older feature requests are still considered to be of value to the community. Thanks!
